### PR TITLE
Implement non-exhaustive enums

### LINKED ISSUE #2524: non-exhaustive enums
A non-exhaustive enum is an enum that may have other unknown members.
This can help create APIs that may be extended in future. This sort of thing frequently comes up in data formats, network protocols, and even C APIs.

Possible syntax:
  - a non-exhaustive enum is created by including `_` as a trailing field.

Semantics: 
  - A non-exhaustive enum must have the size specified: `enum(u5) { X, Y, Z, _ }`
  - If you specify `_` when an enum is already "full" (e.g. if it's `u2` and you specify 4 values followed by a `_`), then there is a compiler error.
  - When `switch`-ed on, an `else` clause is always required
  - `@intToEnum` can never fail with `has no tag matching integer value`
  - `builtin.Enum` gains a new boolean field `is_exhaustive`

Related
  - https://github.com/rust-lang/rfcs/pull/2008
  - https://github.com/rust-lang/rust/pull/51854
  - https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md

### PR DESCRIPTION
**Pull request by Vexu**
Closes #2524

Any suggestions to improve the documentation are welcome.

### CODE REVIEW

File: src/analyze.cpp [hunk 1]
```diff
@@ -2648,6 +2649,29 @@ static Error resolve_enum_zero_bits(CodeGen *g, ZigType *enum_type) {
...
         }
 
+        AstNode *tag_value = field_node->data.struct_field.value;
+
+        if (buf_eql_str(type_enum_field->name, "_")) {
+            if (decl_node->data.container_decl.init_arg_expr == nullptr) {
+                add_node_error(g, field_node, buf_sprintf("non-exhaustive enum must specify size"));
+                enum_type->data.enumeration.resolve_status = ResolveStatusInvalid;
+            }
+            if (log2_u64(field_count - 1) == enum_type->size_in_bits) {
```
Comment (LemonBoy) [inline src/analyze.cpp]: If `field_count` becomes 0 you're going to hit either:
* Undefined behavior with gcc/clang as `log2_u64` calls `clzll` that in turn calls `__builtin_clzll` and that's [undefined](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html) if the parameter is zero
* Garbage with MSVC as `_BitScanReverse` won't write anything in the `lz` variable

I'd say that the `clzll` definition in `util.hpp` needs to be fixed

File: doc/langref.html.in [hunk 2]
```diff
@@ -2893,6 +2893,50 @@ test "switch using enum literals" {
...
+      {#header_open|Non-exhaustive enum#}
+      <p>
+      A Non-exhaustive enum can be created by adding a trailing '_' field.
+      It  must specify a tag type and cannot consume every enumeration value.
+      </p>
+      <p>
+      {#link|@intToEnum#} on a non-exhaustive enum cannot fail.
+      </p>
+      <p>
+      A switch on a non-exhaustive enum can include a '_' prong with the following properties:
```
Comment (daurnimator) [inline doc/langref.html.in]: I don't love the `_` prong of the switch, `else` should be sufficient.

File: src/analyze.cpp [hunk 3]
```diff
@@ -2560,7 +2560,8 @@ static Error resolve_enum_zero_bits(CodeGen *g, ZigType *enum_type) {
 
     assert(!enum_type->data.enumeration.fields);
     uint32_t field_count = (uint32_t)decl_node->data.container_decl.fields.length;
-    if (field_count == 0) {
+    if (field_count == 0 || (field_count == 1 &&
+        buf_eql_str(decl_node->data.container_decl.fields.at(0)->data.struct_field.name, "_"))) {
```
Comment (daurnimator) [inline src/analyze.cpp]: We should allow this. fields often get reserved for later extensions; additionally, an empty non-exhaustive enum was suggested as a clever solution for distinct integer types.

File: src/analyze.cpp [hunk 4]
```diff
@@ -2648,6 +2649,29 @@ static Error resolve_enum_zero_bits(CodeGen *g, ZigType *enum_type) {
...
         }
 
+        AstNode *tag_value = field_node->data.struct_field.value;
+
+        if (buf_eql_str(type_enum_field->name, "_")) {
+            if (decl_node->data.container_decl.init_arg_expr == nullptr) {
+                add_node_error(g, field_node, buf_sprintf("non-exhaustive enum must specify size"));
+                enum_type->data.enumeration.resolve_status = ResolveStatusInvalid;
+            }
+            if (field_count > 1 && log2_u64(field_count - 1) == enum_type->size_in_bits) {
```
Comment (LemonBoy) [inline src/analyze.cpp]: Isn't this impossible to trigger for a non-`extern` enum? The enum tag size is chosen before the `_` field is skipped.
Moreover the `field_count` is now invalid and `enumeration.fields` contains one entry too many.

Comment (Vexu) [reply src/analyze.cpp]: Tag size must be specified for non-exhaustive enums so it is possible to hit.
```zig
const E = enum(u2) {
    a,
    b,
    c,
    d,
    _, // error: non-exhaustive enum specifies every value
};
```

File: doc/langref.html.in (hunk 2 above)
Comment (andrewrk) [reply doc/langref.html.in]: @daurnimator have a look at the rationale in https://github.com/ziglang/zig/issues/2524#issuecomment-570267686. The `_` is admittedly introducing language complication, however, it's required to be part of non-exhaustive enums in order to not give up the ability to have compile errors for missing enum values.

`_` vs `else` is the difference between API compatibility and ABI compatibility. For non-exhaustive enums, both `_` and `else` make the code ABI compatible. However `_` means the code is ABI compatible but not API compatible with new tags. `_` means you get a compile error when a new tag is added, but you will not get *undefined behavior at runtime* when a tag is added.

**Pull request merged**